### ISSUE TITLE
Warning message when saving model 

### ISSUE THREAD

**Issue by ivan-marroquin** [dmlc/xgboost#6060]
Hi there, 

I am using Python 3.6.5 on a windows machine, with xgboost 1.1.1 and scikit-learn 0.20.2.

So, I train a classifier defined as follows:

best_trained_model= xgb.XGBClassifier(num_class= n_class, max_depth= best_max_depth, n_estimators= 100, verbosity= 0, objective= 'multi:softmax', booster= 'gbtree', tree_method= 'exact', n_jobs= cpu_count, learning_rate= best_learning_rate, gamma= 0.25,  reg_alpha= best_reg_alpha, reg_lambda= best_reg_lambda, random_state= 1969)

for which, the value for parameters "max_depth", "learning_rate", "reg_alpha" and "reg_lambda" were optimized using automated machine learning. The training part worked just fine.

However, when I use the following command to save the trained model: 

best_trained_model.save_model(train_model_path + model_fn)

I get warning messages about "max_depth", "learning_rate", "reg_alpha" and "reg_lambda" are not saved in Scikit-Learn meta. What does this imply? What happens if I want to reuse the trained model?

With xgboost 0.90 and using the same Python code, I did not have such warning message.

Thanks for your help,

Ivan

**Comment by trivialfis** [comment]
> "max_depth", "learning_rate", "reg_alpha" and "reg_lambda" were optimized using automated machine learning.

My guess is these objects can not be serialized by Python JSON module.  (It's part of Python standard library).

> What does this imply? What happens if I want to reuse the trained model?

Nothing.  Training parameters are not supposed to be saved in a normal sense.  See https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html for details about what should be saved in a model.  In short, if you use the model for prediction, these parameters are useless.  If you want to add more trees to your model, you need to provide a new set of parameters by `set_params` before calling `fit`.

> With xgboost 0.90 and using the same Python code, I did not have such warning message.

We didn't throw it back then.

**Comment by ivan-marroquin** [comment]
Hi @trivialfis 

Many thanks for your prompt answer. 

However, I would like to draw your attention to the following points. Having the ability to store hyperparameters and their values is very important. Think about reproducibility and accountability. Without this information, how can we track back our steps? or understand, how someone built his classifier/regressor model? or even, what hyperparameter values are more important than others?

I am hoping that the developers of xgboost will add the functionality to store hyperparameter values, as it was in xgboost version 0.90

Thanks for such wonderful package!

Ivan

**Comment by hcho3** [comment]
@ivan-marroquin In fact, you can retrieve the list of hyperparameters by calling the `save_config()` method.

**Comment by trivialfis** [comment]
As mentioned, you can call save config.  If stability is not of concern, pickle is another option.  See the doc referenced in previous comment.
